Some more reader refactoring

fingolfin opened the pull request:
This moves the documentation comment for `ReadCallVarAss` to the right place; and moves up the definition of `findValueInNams`, so that we can use it inside `ReadCallVarAss`.

ChrisJefferson commented:
I only have one comment -- "There seems to be no valid justification for that (other
than that code using more than 1024 entries in 'nams' probably is seriously
crazy to start with)."

I assume at some point someone thought there was a sensible reason for that. Perhaps make a test which triggers this case: Nest some functions with lots of varaibles, access variables in parents, check that there isn't something else hidden we are unaware of?

fingolfin commented:
Based on the code formatting, I am pretty sure that someone was @stevelinton ... So I just did what I should have done before, and tracked this through the full (and sadly private :/) repository history. Turns out, this commit introduced the checks for `indx`:
```
commit 6ee3220bf57b20057ef8bd0f107110ec69f1c38f
Author: Steve Linton <steve.linton@st-andrews.ac.uk>
Date:   1998-01-05 17:24:03 +0000

    Kernel updates from my holidays

    1. Save/load now works OK even with the compiler (as far as I can tell)
    2. "quit;" leaves any number of nested reads, "QUIT;" always leaves GAP
    3. OnBreak() is executed just before entering a break loop, default
       action is to call Where()
    4. scope in a break loop now includes calling routines (dynamic) as
       well as routines enclosing the current function definition (static)
    5. I may have fixed the Alpha/weakpointer bug

        Steve
```

Looking at the code reveals that back then, only 10 bits were used to store the local name index, because indeed three values were crammed into a single 32bit integer: nest0, nest, and indx. Here's a code snippet introduced by the above commit:
```C
   indx = LEN_LIST( nams );
   if (indx >= 1024)
     {
       Pr("Warning; Ignoring local names after 1024th in search for %s\n",
          (Int) Value,
          0L);
       indx = 1023;
     }
        for ( ; 1 <= indx; indx-- ) {
     if ( SyStrcmp( Value, CSTR_STRING(ELM_LIST(nams,indx)) ) == 0 ) {
       type = 'd';

       /* Ultrix 4.2 cc get's confused if the UInt is missing     */
       var = ((UInt) nest0 << 20) + ((UInt)nest << 10) + indx;
       break;
     }
```
Note that back then, we also limited the value of `nest` to 1024=2^10 and the value of `nest0` to 4096=2^12, consistent with my explanation.

Then later this was improved, by realizing that we don't need to cram `nest0` in there:
```
commit 13d9b11eb1ccdb06123afd0b1a1efe308d137f30
Author: Burkhard Höfling <burkhard@hoefling.name>
Date:   2005-10-17 09:31:34 +0000

    Enabled access to local variables anywhere on the stack from the break loop. BH
```
which contains, among others, these changes:
```patch
...
--- a/src/read.c
+++ b/src/read.c
@@ -313,14 +313,14 @@ void ReadCallVarAss (
        type = 'd';

        /* Ultrix 4.2 cc get's confused if the UInt is missing     */
-       var = ((UInt) nest0 << 20) + ((UInt)nest << 10) + indx;
+       var = ((UInt)nest << 16) + indx;
        break;
          }
        }
         }
         lvars = ENVI_FUNC( PTR_BAG( lvars )[0] );
         nest++;
-   if (nest >= 1024)
+   if (nest >= 65536)
      {
        Pr("Warning: abandoning search for %s at 1024th higher frame\n",
           (Int)Value,0L);
@@ -329,7 +329,7 @@ void ReadCallVarAss (
       }
       lvars0 = PTR_BAG( lvars0 )[2];
       nest0++;
-   if (nest0 >= 4096)
+   if (nest0 >= 65536)
      {
        Pr("Warning: abandoning search for %s 4096 frames up stack\n",
           (Int)Value,0L);
...
```
So the limits for nest and nest0 were increased to 2^16, but for some reasons, that for `indx` was not -- looks like a simple oversight to me. Also, the limit for `nest` to be 2^16 obviously assumes 32bit integers, with 64bit integers it'd be 2^48. And finally, there actually is no reason to limit `nest0` here, because it is passed (and used) as an `UInt`, so can use the full 32/64 bits.

Thus, *if* we want to perform bounds checks here, then the limits should be `indx < 2^16` and (only on 32 bit systems) `nest < 2^16`, and no bounds check on `nest0`. Now, a value of `nest >= 2^16` would mean that somebody nested function definitions so deep that a higher var is 2^16 levels up. So for example. they wrote a function like the following, only instead of nesting 4 levels (depending on how you count), they nested it > 2^16 levels deep:
```gap
f := function()
  local x1; x1:=1;
  return function()
    local x2; x2:=2;
    return function()
      local x3; x3:=3;
       return function()
          return 1/0; # trigger an error so that we end up in an error.
       end;
     end;
   end;
 end;
```
Of course I can write a program now that generates a function like that. And I think it'd trigger the check for `nest`. But it seems rather crazy to have such code.

But now note that the far more common case is to access lvars and hvars, and *that* code never performed any bounds checks. So it seems weird to check for bounds in the code used for accessing vars in the break loop only...

So I guess also this code should check bounds (I am quoting the revised form found in this PR, but this remark also applies to it before this PR):
```C
    const UInt countNams = LEN_PLIST(STATE(StackNams));
    for (nest = 0; nest < countNams; nest++) {
        nams = ELM_LIST(STATE(StackNams), countNams - nest);
        indx = findValueInNams(nams, 1, LEN_PLIST(nams));
        if (indx != 0) {
            type = (nest == 0) ? 'l' : 'h';
            var = (nest << 16) + indx;
            break;
        }
    }
```

Now, instead of actually checking here, I guess we ought to check 
* whenever we `PushPlist(STATE(StackNams, ...))` let's check for an "overflow" and error out,
* whenever reading a function in, error out if its `nams` exceeds 1024.

Then we can drop the checks in `ReadCallVarAss`, resp. replace them with `GAP_ASSERT` checks (purely out of paranoia)

stevelinton commented:
Thanks for the careful analysis. It fits entirely with what I can recall at this distance, and what you propose sounds entirely reasonable. If anyone ever starts machine-generating GAP code, 1024 local variables could possibly be a limitation, but I don't see how anyone could ever want to lexically nest functions more than a few levels deep.

fingolfin commented:
@stevelinton yeah, that's what I think, too. And luckily, the actual limit on locals right now is 65536 not 1024 (at least with the single check for 1024 removed, as in this PR).

ChrisJefferson commented:
@fingolfin Do you think this is now fine for merging (I'm happy!) -- just wanted to check you didn't want to make any changes based on the discussion.

fingolfin commented:
I have updated this now based on our discussion (test are running). I added a check which will cause GAP to reject functions with >= 65536 arguments and locals. I also wrote a program which generates functions that have 65535 arguments (and then manually edited them to exceed the number of arguments+locals, by either adding another argument or a local). The check works, but (a) the test file is ~250 KB and (b) GAP takes about 10 seconds to parse it. The latter is caused by the naive linear search we perform to lookup symbol names (i.e. in `findValueInNams`). While we could speed that up using a hash table (e.g. by (ab)using a precord for that), I don't think it's worth the effort, because quite frankly if you write such code, then you deserve to be punished ;-). Also, once the parsing and coding is done, no such lookup happens anymore.

So instead of trying to speed up that case, we might want to instead artificially restrict GAP to 1024 or so locals (parsing those is still very fast). But that could happen in a later PR, if we wanted to do it.
